src/engine/DebugOverlays.cs

```diff
@@ -0,0 +1,213 @@
+        if (performanceMetrics.Visible)
+        {
+            fpsLabel.Text = new LocalizedString("FPS", Engine.GetFramesPerSecond()).ToString();
+            deltaLabel.Text = new LocalizedString("FRAME_DURATION", delta).ToString();
+
+            var currentProcess = Process.GetCurrentProcess();
+
+            var processorTime = currentProcess.TotalProcessorTime;
+            var threads = currentProcess.Threads.Count;
+            var usedMemory = Math.Round(currentProcess.WorkingSet64 / (double)Constants.MEBIBYTE, 1);
+
+            // These don't seem to work:
+            // Performance.GetMonitor(Performance.Monitor.Physics3dActiveObjects),
+            // Performance.GetMonitor(Performance.Monitor.Physics3dCollisionPairs),
+            // Performance.GetMonitor(Performance.Monitor.Physics3dIslandCount),
+
+            metricsText.Text =
+                new LocalizedString("METRICS_CONTENT", Performance.GetMonitor(Performance.Monitor.TimeProcess),
+                        Performance.GetMonitor(Performance.Monitor.TimePhysicsProcess),
+                        entities, children, spawnHistory.Sum(), despawnHistory.Sum(),
+                        Performance.GetMonitor(Performance.Monitor.ObjectNodeCount), usedMemory,
+                        Math.Round(Performance.GetMonitor(Performance.Monitor.RenderVideoMemUsed) / Constants.MEBIBYTE,
+                            1),
+                        Performance.GetMonitor(Performance.Monitor.RenderObjectsInFrame),
+                        Performance.GetMonitor(Performance.Monitor.RenderDrawCallsInFrame),
+                        Performance.GetMonitor(Performance.Monitor.Render2dDrawCallsInFrame),
+                        Performance.GetMonitor(Performance.Monitor.RenderVerticesInFrame),
+                        Performance.GetMonitor(Performance.Monitor.RenderMaterialChangesInFrame),
+                        Performance.GetMonitor(Performance.Monitor.RenderShaderChangesInFrame),
+                        Performance.GetMonitor(Performance.Monitor.ObjectOrphanNodeCount),
+                        Performance.GetMonitor(Performance.Monitor.AudioOutputLatency) * 1000, threads, processorTime)
+                    .ToString();
+
+            entities = 0;
+            children = 0;
+
+            spawnHistory.AddToBack(currentSpawned);
+            despawnHistory.AddToBack(currentDespawned);
+
+            while (spawnHistory.Count > SpawnHistoryLength)
+                spawnHistory.RemoveFromFront();
+
+            while (despawnHistory.Count > SpawnHistoryLength)
+                despawnHistory.RemoveFromFront();
+
+            currentSpawned = 0;
+            currentDespawned = 0;
+        }
+
+        // FPS counter
+        if (fpsCounter.Visible)
+            fpsDisplayLabel.Text = new LocalizedString("FPS", Engine.GetFramesPerSecond()).ToString();
+    }
+
+    [RunOnKeyDown("toggle_metrics", OnlyUnhandled = false)]
+    public void OnPerformanceMetricsToggled()
```

Comment:
Is it possible to move overlay-specific callbacks to the partial classes?

Reply:
This function is actually DebugPanel (the dialog) callback, which is in this file.

Reply:
Oh right, I failed to notice that. This is good to go then.